ENH: Interactive Volatility Control Charts

thermokarst opened the pull request:
Fixes #79 (well, all but zooming, will open a new issue for that)
Fixes #75 ~(need to double-check this, I think the mouseover only works for the mean lines ATM)~
Fixes #82 
Fixes #83
Fixes #90 (?)

---

This PR is a bit large, so I broke it up into sequential commits that build on each other, and will hopefully make review a bit more digestible. Whenever it gets merged, lets squash it all down into one commit. Thanks!

---

If this vega business works out, we should move a lot of this boilerplate over to q2templates. Lets open an issue (if necessary) when this PR is merged.

---

cc @ElDeveloper

thermokarst committed:
ENH: Interactive volatility viz

thermokarst committed:
SQUASH: drop original volatility code

thermokarst committed:
SQUASH: rename volatility2 -> volatility

thermokarst committed:
SQUASH: Fixing up existing tests, refact. vldtn.

thermokarst committed:
SQUASH: Add a few new tests

thermokarst committed:
SQUASH: Add mouseover to spaghetti lines

thermokarst committed:
SQUASH: remove dead code that was left behind

nbokulich commented on q2_longitudinal/_longitudinal.py:
```diff
@@ -162,41 +168,66 @@ def linear_mixed_effects(output_dir: str, metadata: qiime2.Metadata,
-    _visualize(output_dir, plot=chart, summary=summary, raw_data=raw_data,
-               plot_name='Control charts')
+def volatility(output_dir: str, metadata: qiime2.Metadata,
+               default_group_column: str, default_metric: str,
```
perhaps the `default_*` columns should be optional?

in general, clearing these out would simplify the interface, but I do like setting these defaults (saves a few clicks!). Making these optional may be a good compromise.

nbokulich commented on q2_longitudinal/assets/volatility/index.html:
```diff
@@ -0,0 +1,155 @@
+    cleanupEl('#toggle-error-bars', 'Show error bars');
+    cleanupEl('#toggle-global-mean', 'Show global mean');
+    cleanupEl('#toggle-global-control-limits', 'Show global control limits');
+    $('#toggle-global-control-limits').append('<p>+/- 2x and 3x standard deviations</p>');
```
should we clarify that these are 2x and 3x standard deviations from global mean?

nbokulich commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -212,39 +211,36 @@
-    input_descriptions={'table': (
-        'Feature table to optionally use for paired comparisons.')},
+    inputs={
+        'table': FeatureTable[RelativeFrequency],
```
This is where interactive metric selection could get tricky... using a feature table could lead to a massive number of possible metrics to choose from. This will not only be messy on the interface, but might cause this to lag/crash for some users? I will test this out to try and break it.

A few ideas:
1. The most basic — but unsatisfying — solution might be to make "metric" a required parameter and not allow metric selection within the viz if only one metric can be input.
2. Same as above, but accept a list of metrics on input. These would be selectable. That would be nice — there will often be a few metrics that a user would like to plot.
3. Same as 2, but make it optional? i.e., allow selection of all numeric columns as metrics if no list is passed. This is nice because one really nice thing about interactive selection is that you don't need to know the name of your metric (some metric names can be non-intuitive, e.g., pielou's E instead of evenness).

nbokulich replied:
CORRECTION it seems that this may be broken — when I input a feature table, none of that data is available in the viz.

thermokarst replied:
Just tested this with the ECAM table you provided in an out-of-band message (after fixing up the broken table handling) - I don't think these column will be too painful for performance. I like the idea of having an optional list of columns to include in the viz, but how about for now we just point people to `feature-table filter-features` if they run up against a performance issue? Would that work for you?

nbokulich replied:
that works for me — I can also document this in the tutorial

ECAM is not massive, but it's a pretty good size (several hundred samples and SVs?) so is probably a reasonable test case — if it's not causing problems, then I think we are okay for now.

nbokulich commented on q2_longitudinal/_utilities.py:
```diff
@@ -404,130 +402,6 @@ def _regplot_subplots_from_dataframe(state_column, metric, metadata,
-    return ax
-
-
-def _set_xtick_interval(xtick_interval, states):
```
This is not a feature that we want to strip away (though there must be a better way to do this in vega rather than keeping this function). 

See the broken qzv I've attached to the review to see why... we either need to support explicit interval setting or add sensible auto-scaling in vega.

nbokulich commented on q2_longitudinal/_vega.py:
```diff
@@ -0,0 +1,811 @@
+                },
+            },
+            {
+                'name': 'meanLineThickness',
```
can we also add a slider bar for spaghetti thickness?

nbokulich commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -362,71 +329,41 @@ def test_linear_mixed_effects_taxa(self):
-            'time', 'metric', plot_this, None, ci=95, palette='Set1', ax=None,
-            legend=True, color=None, xtick_interval=None)
-
-    def test_volatility_table_data(self):
```
since table inputs are currently breaking, maybe we should add a test for this? I appreciate that testing visualizations is a PITA so maybe don't test at the action level, just test that feature data is actually being merged and preserved in the metadata, such that it should be accessible in a qzv?

nbokulich replied:
an easy test might be to just use a feature ID as a `default_metric` since the validation comes after merging. Unless if features are being dropped later, e.g., by vega.

thermokarst committed:
SQUASH: Addressing @nbokulich's review comments

- Made the default_* parameters optional. If they aren't provided, the
  first column in the respective group is selected.
- Refactored table/metadata merging so that it now works
  (:crossed-fingers:)
- Updated the text on the control limit widget to clarify global mean
  basis.
- Fixed x-axis sorting.
- Applied "nice" x-axis formatting rules.
- Slider widget added for :spaghetti: line thickness.
- Reordered some of the interactive form elements
- Replaced :spaghetti: visibility toggle with an opacity slider widget.
- Replaced mean-line visibility toggle with an opacity slider widget.
- Add sample ID as available "group-by" column.
- Added additional tests.

nbokulich reviewed (changes requested):
BEAUTIFUL.

I am loving vega, and you have done such an awesome job.

There are a few major gaps, though. I'll summarize here but more detail is given in-line:
1. Table inputs do not appear to merge correctly with the metadata, and hence cannot be used as metrics. 
2. x-tick sorting is broken (we discussed this)
3. x-tick intervals should either be explicitly set or sensibly auto-scaled. This was a feature we added to the static viz due to practical needs...

There are a few other minor features that I think we will need on release. These are mentioned in-line.

thermokarst committed:
Merge branch 'master' into spaghettwo

* master:
  ENH: added citations (#87)
  ENH: group_column is now optional (#81)

nbokulich commented on q2_longitudinal/_longitudinal.py:
```diff
@@ -207,15 +217,10 @@ def volatility(output_dir: str, metadata: qiime2.Metadata,
-
-    group_columns = list(categorical.columns.keys())
+    # We want to include the original Sample ID in the viz, so reset the index.
+    data = metadata.to_dataframe().reset_index(drop=False)
```
it looks like this adds the sample name, which does not let us track individual spaghetti (sample name corresponds to individual points).

we want `individual_id_column` — which may well be numeric (as it is with ECAM) and hence may not appear in the group category options by default.

thermokarst replied:
Ah, thanks for the clarification - I completely misunderstood #83. Should I keep this change, in addition to the `individual_id_column`?

nbokulich replied:
My fault, #83 does not have enough description. The goal is to calculate mean lines across all samples, rather than splitting by group. Your change does 50% of that — there is no group, but there is no mean line, either. So we need a way to group all samples together for calculating a "group" mean, perhaps by generating a new metadata column (though that risks overwriting an existing column that happens to be named something like "NA").

For some reason I thought that this change was actually targeting #90 — with the `individual_id_column` change added in you can claim that issue too! or not, if you just want to get this off your plate — that was not in the original manifest for this PR so I would understand.

nbokulich reviewed (changes requested):
Thanks @thermokarst ! I am loving these new changes! (and learning so much about what `Metadata` can do! 😉 )

My comments at this point are mostly at the interface level (I will send an example qzv offline):
1. It looks like time point 0 is being dropped for some reason — could you investigate?
2. To select individual spaghetti, we need `individual_id_column` (not sample ID) to be forced into the group columns — see comment inline
3. The plot legend cuts off labels — for an example, try setting group column to "id" in the example I'm sending. This is a problem both for number of labels and label length. Number of labels we should fix in this PR if possible, but there might not be a good compromise for label length (we don't really want to wrap long labels so cutting off very long labels might be all we can do. In any case, this is a bug we can raise an issue for now and fix later)
4. `individual_id_column` should be added to the mouse-over metadata shown for 🍝

thermokarst committed:
SQUASH: addressing @nbokulich's comments

thermokarst committed:
SQUASH: attempting to address @gregcaporaso review

thermokarst committed:
SQUASH: Make mouseovers easier with symbol ctrls

thermokarst committed:
SQUASH: missing package_data
